core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -150,6 +166,110 @@ public Object visitEvictCommand(InvocationContext ctx, EvictCommand command) thr
+            transport.getAddress());
+      Map<Object, Object> values = (Map<Object, Object>) invokeNextInterceptor(ctx, command);
+      
+      if (ctx.isOriginLocal()) {
+         CacheTopology afterTopology = stateTransferManager.getCacheTopology();
+         if (beginTopology.getTopologyId() != afterTopology.getTopologyId()) {
+            Map<Object, ?> remotelyRetrieved = command.getRemotelyFetched();
+            // Locally we have to see if anything is null, this could have meant we had
+            // a miss right when a topology changed - so find them and retry!
+            Collection<?> originalKeys = command.getKeys();
+            List<Object> keysToTryAgain = new ArrayList<Object>(originalKeys.size());
+            for (Object key : originalKeys) {
+               if (values.containsKey(key)) {
+                  Object value = values.get(key);
+                  if (value == null && !remotelyRetrieved.containsKey(key)) {
+                     // If the value was returned as null and it wasn't a remote lookup
+                     // that means it was a possible rehash miss, so we have to look
+                     // it up again
+                     keysToTryAgain.add(key);
+                  }
+               } else {
+                  // This occurs if a remote miss had a rehash while processing
+                  keysToTryAgain.add(key);
+               }
+            }
+            if (!keysToTryAgain.isEmpty()) {
+               try {
+                  log.tracef("Retrying keys %s", keysToTryAgain);
+                  command.setKeys(keysToTryAgain);
+                  values.putAll((Map<Object, Object>) visitGetAllCommand(ctx, command));
+               } finally {
+                  command.setKeys(originalKeys);
+               }
+            }
+         } else {
+            Collection<?> originalKeys = command.getKeys();
+            int missingKeys;
+            // If these don't match it means that a remote node saw a new topology and
+            // didn't have the value but our local node hasn't yet seen the new topology
+            if ((missingKeys = originalKeys.size() - values.size()) > 0) {
+               List<Object> keysToTryAgain = new ArrayList<Object>(missingKeys);
+               for (Object key : originalKeys) {
+                  if (!values.containsKey(key)) {
+                     keysToTryAgain.add(key);
+                  }
+               }
+               if (!keysToTryAgain.isEmpty()) {
+                  try {
+                     log.infof("Retrying keys %s from stable topology", keysToTryAgain);
+                     command.setKeys(keysToTryAgain);
+                     // Add in a random sleep here just to reduce the chance of getting
+                     // StackOverflow if it takes too long for a suspected node to be
+                     // removed from the view
+                     Thread.sleep(10);
```

Comment:
Indeed, for non-tx write commands we expect a new cache topology when we get a SuspectException. I'm trying to fix the just-suspect-not-actually-leaving exceptions for ISPN-4546.

I'm not sure it's ok to rely on the topology ids here to decide whether to retry or not, though, I'd rather just ignore the topology id and check whether we got a value for all the keys we requested.